Internal: Move components data to store [ED-20934]

MuliElementor opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

<!--start_gitstream_placeholder-->
### ✨ PR Description
Purpose: Move component data management from React Query to Redux Store for centralized state management and improved data flow.

Main changes:
- Implemented Redux store slice for components with load/create actions and status tracking
- Replaced React Query hooks with Redux selectors and thunks for data fetching operations
- Added PopulateStore component to initialize store data on application startup
- Exported CreateComponentPayload type for better API integration

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using. **[We'd love your feedback!](mailto:product@linearb.io)** 🚀</sub>
<!--end_gitstream_placeholder-->

MuliElementor committed:
Internal: Enhance component management with new update functionality and store integration

MuliElementor committed:
add put  endpoint

MuliElementor committed:
Update dependencies and enhance component tests

- Added "@elementor/store" dependency version 3.33.0 to package.json.
- Refactored component tests to utilize new testing utilities and improve structure.
- Enhanced test coverage for ComponentsList and ComponentItem, including loading states and search functionality.
- Updated mock implementations for better isolation in tests.

MuliElementor committed:
update

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
Refactor component store integration and loading logic

- Updated PopulateStore to utilize loadComponents for data fetching.
- Enhanced store slice to manage loading states with extraReducers.
- Introduced useLoadStatus hook for better loading state management in components.
- Improved component tests to reflect new store structure and loading behavior.

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

# Conflicts:
#	packages/package-lock.json

MuliElementor committed:
update

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

IshayMaya commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -1,246 +1,226 @@
-		expect( screen.getByLabelText( 'More actions' ) ).toBeInTheDocument();
-	} );
+			// Assert
+			expect( screen.getByText( 'Text that explains that there are no Components yet.' ) ).toBeInTheDocument();
```
it doesn't test the components

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
update components

MuliElementor committed:
Refactor component management: Moved loadComponents to actions, updated state management for createComponent, and improved error handling in CreateComponentForm. Removed obsolete load-components file.

MuliElementor committed:
update

MuliElementor committed:
Enhance CreateComponentForm tests: Adjusted timeout for async operations and improved assertions for loading states. Removed obsolete comments for clarity.

elementorbot committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

IshayMaya commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -209,18 +228,103 @@ describe( 'ComponentsTab', () => {
+			expect( screen.getByText( 'Try something else.' ) ).toBeInTheDocument();
+		} );
+
+		it( 'should handle undefined components gracefully', () => {
```
already tested

IshayMaya commented on packages/packages/core/editor-components/src/hooks/use-create-component.ts:
```diff
@@ -1,14 +1,22 @@
+	const createStatus = useSelector( selectCreateStatus );
+
+	const createComponentAction = async ( payload: CreateComponentPayload ) => {
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
can't we infer the actual type?
where is the `payload` in `result` come from?

IshayMaya commented on packages/packages/core/editor-components/src/hooks/use-create-component.ts:
```diff
@@ -1,14 +1,22 @@
+	const createComponentAction = async ( payload: CreateComponentPayload ) => {
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		const result = await dispatch( createComponent( payload ) as any );
+		return result.payload;
```
we are not fetching new data from the server?

MuliElementor committed:
Refactor actions: Consolidated createComponent and loadComponents imports into a single actions module, removed obsolete action files, and updated related components and hooks accordingly.

MuliElementor committed:
update

MuliElementor committed:
Refactor import statement: Updated loadComponents import to use named import from actions module for improved clarity and consistency.

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor commented on packages/packages/core/editor-components/src/actions.ts:
```diff
@@ -0,0 +1,18 @@
+import { __createAsyncThunk as createAsyncThunk } from '@elementor/store';
```
changed

MuliElementor committed:
Refactor: Move component actions to thunks and update related imports

- Deleted the actions.ts file and moved createComponent and loadComponents to thunks.
- Updated imports in populate-store.ts, store.ts, and test files to reflect the new thunks structure.
- Refactored hooks and selectors to improve loading state management for components.
- Cleaned up unused test cases in components-tab.test.tsx and create-component-form.test.tsx.

MuliElementor committed:
Refactor: Simplify component mapping in ComponentsList

- Updated the mapping of components in the ComponentsList component to use a more concise syntax.
- Removed optional chaining for component IDs to ensure cleaner code.

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MiryamOren commented on packages/packages/core/editor-components/src/store.ts:
```diff
@@ -0,0 +1,73 @@
+
+export const selectComponents = createSelector( selectData, ( data: Component[] ) => data );
+export const selectLoadIsPending = createSelector( selectLoadStatus, ( status ) => status === 'pending' );
+export const selectLoadError = createSelector( selectLoadStatus, ( status ) => status === 'error' );
```
To be consistent with `selectLoadIsPending`
```suggestion
export const selectLoadIsError = createSelector( selectLoadStatus, ( status ) => status === 'error' );
```

MiryamOren commented on packages/packages/core/editor-components/src/store.ts:
```diff
@@ -0,0 +1,73 @@
+export const selectLoadIsPending = createSelector( selectLoadStatus, ( status ) => status === 'pending' );
+export const selectLoadError = createSelector( selectLoadStatus, ( status ) => status === 'error' );
+export const selectCreateIsPending = createSelector( selectCreateStatus, ( status ) => status === 'pending' );
+export const selectCreateError = createSelector( selectCreateStatus, ( status ) => status === 'error' );
```
Same here
```suggestion
export const selectCreateIsError = createSelector( selectCreateStatus, ( status ) => status === 'error' );
```

MiryamOren commented on packages/packages/core/editor-components/src/components/components-tab/components-list.tsx:
```diff
@@ -126,12 +128,12 @@ const EmptySearchResult = () => {
-		components: components?.filter( ( component ) =>
-			component.name.toLowerCase().includes( searchValue.toLowerCase() )
+		components: components?.filter(
+			( component ) => component?.name?.toLowerCase().includes( searchValue.toLowerCase() )
```
```suggestion
			( component ) => component.name.toLowerCase().includes( searchValue.toLowerCase() )
```

MiryamOren commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -1,246 +1,282 @@
-		expect( screen.getByText( 'Test Component' ) ).toBeInTheDocument();
-		expect( screen.getByLabelText( 'More actions' ) ).toBeInTheDocument();
-	} );
+		it( 'should render components list with multiple components', () => {
```
This one can be merged with the previous test

MiryamOren commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -1,246 +1,282 @@
+			expect( screen.getByText( 'Delete' ) ).toBeInTheDocument();
+		} );
+
+		it( 'should render draggable component item', () => {
```
Am I missing something? This test still doesn't check darggability

MiryamOren commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -1,246 +1,282 @@
+			).not.toBeInTheDocument();
+		} );
+
+		it( 'should filter components based on search value', () => {
```
This test is included in `should handle search functionality and show filtered results`, I think we can delete it

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
Refactor: Rename error selectors for consistency in store

- Updated selector names from `selectLoadError` to `selectLoadIsError` and from `selectCreateError` to `selectCreateIsError` for improved clarity.
- Adjusted related imports in `use-create-component` hook.
- Cleaned up test file by removing redundant mock declarations and unnecessary test cases, while enhancing existing tests for better readability and maintainability.

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MiryamOren commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -1,246 +1,248 @@
+			const componentItem = screen.getByRole( 'button', { name: /Button Component/ } );
+			expect( componentItem ).toBeInTheDocument();
+			expect( componentItem ).toHaveAttribute( 'draggable', 'true' );
+			expect( screen.getByLabelText( 'More actions' ) ).toBeInTheDocument();
```
We are checking it in a different test
```suggestion
```

MiryamOren commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -1,246 +1,248 @@
+			expect( componentItem ).toBeInTheDocument();
+			expect( componentItem ).toHaveAttribute( 'draggable', 'true' );
+			expect( screen.getByLabelText( 'More actions' ) ).toBeInTheDocument();
+			expect( screen.getByText( 'Button Component' ) ).toBeInTheDocument();
```
Already checked in line 167
```suggestion
```

MiryamOren commented on packages/packages/core/editor-components/src/components/__tests__/components-tab.test.tsx:
```diff
@@ -17,12 +24,6 @@ import { ComponentItem } from '../components-tab/components-item';
 import { ComponentsList } from '../components-tab/components-list';
 import { SearchProvider } from '../components-tab/search-provider';
 
-// Mock external dependencies
```
Why did you move it? mocks should be under the imports

MiryamOren commented on packages/packages/core/editor-components/src/hooks/use-create-component.ts:
```diff
@@ -1,13 +1,22 @@
+export const useCreateComponent = () => {
+	const dispatch = useDispatch();
+	const isPending = useSelector( selectCreateIsPending );
+	const error = useSelector( selectCreateIsError );
```
```suggestion
	const isError = useSelector( selectCreateIsError );
```

MuliElementor committed:
Refactor: Update error handling in useCreateComponent hook and clean up tests

- Renamed `error` to `isError` in the `useCreateComponent` hook for consistency with other selectors.
- Removed redundant mock declarations and unnecessary test cases in `components-tab.test.tsx`, improving test clarity and maintainability.

MuliElementor committed:
Merge branch 'main' into ED-20934-move-components-data-to-store

MuliElementor committed:
update
